azurelinuxagent/ga/extension_telemetry.py

```diff
@@ -0,0 +1,402 @@
+                return events_list
+
+            for extension_handler_with_event_dir in extension_handler_with_event_dirs:
+                handler_name = extension_handler_with_event_dir[0]
+                handler_event_dir_path = extension_handler_with_event_dir[1]
+                self._capture_extension_events(handler_name, handler_event_dir_path, events_list)
+        except Exception as e:
+            msg = "Unknown error occurred when trying to collect extension events. Error: {0}".format(ustr(e))
+            add_event(op=WALAEventOperation.ExtensionTelemetryEventProcessing, message=msg, is_success=False)
+        finally:
+            # Always ensure that the events directory are being deleted each run,
+            # even if we run into an error and dont process them this run.
+            self._ensure_all_events_directories_empty(extension_handler_with_event_dirs)
+            return events_list
+
+    @staticmethod
+    def _get_extension_events_dir_with_handler_name(extension_log_dir):
+        """
+        Get the full path to events directory for all extension handlers that have one
+        :param extension_log_dir: Base log directory for all extensions
+        :return: A list of full paths of existing events directory for all handlers
+        """
+        extension_handler_with_event_dirs = []
+
+        for ext_handler_name in os.listdir(extension_log_dir):
+            # Check if its an Extension directory
+            if not os.path.isdir(os.path.join(extension_log_dir, ext_handler_name)) \
+                    or re.match(HANDLER_NAME_PATTERN, ext_handler_name) is None:
+                continue
+
+            # Check if EVENTS_DIRECTORY directory exists
+            extension_event_dir = os.path.join(extension_log_dir, ext_handler_name, EVENTS_DIRECTORY)
+            if os.path.exists(extension_event_dir):
+                extension_handler_with_event_dirs.append((ext_handler_name, extension_event_dir))
+
+        return extension_handler_with_event_dirs
+
+    def _capture_extension_events(self, handler_name, handler_event_dir_path, events_list):
+        """
+        Capture Extension events and add them to the events_list
+        :param handler_name: Complete Handler Name. Eg: Microsoft.CPlat.Core.RunCommandLinux
+        :param handler_event_dir_path: Full path. Eg: '/var/log/azure/Microsoft.CPlat.Core.RunCommandLinux/events'
+        :param events_list: List of captured extension events
+        """
+
+        convert_to_mb = lambda x: (1.0 * x)/(1000 * 1000)
+
+        # Filter out the files that do not follow the pre-defined EXTENSION_EVENT_FILE_NAME_REGEX
+        event_files = [event_file for event_file in os.listdir(handler_event_dir_path) if
+                       re.match(self._EXTENSION_EVENT_FILE_NAME_REGEX, event_file) is not None]
+        # Pick the latest files first, we'll discard older events if len(events) > MAX_EVENT_COUNT
+        event_files.sort(reverse=True)
```

Comment:
Is the sorting key guaranteed to be the last modified time? Or is it based on the file name, which is epoch time?

Reply:
The sorting is based on the file name which is supposed to be epoch time.